TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java

```diff
@@ -0,0 +1,90 @@
+package org.firstinspires.ftc.teamcode.util;
+
+import com.acmerobotics.roadrunner.util.NanoClock;
+import com.qualcomm.robotcore.hardware.DcMotorEx;
+
+/**
+ * Wraps a motor instance to provide corrected velocity counts and allow reversing without changing the corresponding
+ * slot's motor direction
+ */
+public class Encoder {
+    public static int CPS_STEP = 0x10000;
+
+    public static double inverseOverflow(double input, double estimate) {
+        double real = input;
+        while (Math.abs(estimate - real) > CPS_STEP / 2.0) {
+            real += Math.signum(estimate - real) * CPS_STEP;
+        }
+        return real;
+    }
+
+    public enum Direction {
+        FORWARD(1),
+        REVERSE(-1);
+
+        private int multiplier;
+
+        Direction(int multiplier) {
+            this.multiplier = multiplier;
+        }
+
+        public int getMultiplier() {
+            return multiplier;
+        }
+    }
+
+    private DcMotorEx motor;
+    private NanoClock clock;
+
+    private Direction direction;
+
+    private int lastPosition;
+    private double velocityEstimate;
+    private double lastUpdateTime;
+
+    public Encoder(DcMotorEx motor, NanoClock clock) {
+        this.motor = motor;
+        this.clock = clock;
+
+        this.direction = Direction.FORWARD;
+
+        this.lastPosition = 0;
+        this.velocityEstimate = 0.0;
+        this.lastUpdateTime = clock.seconds();
+    }
+
+    public Encoder(DcMotorEx motor) {
+        this(motor, NanoClock.system());
+    }
+
+    public Direction getDirection() {
+        return direction;
+    }
+
+    /**
+     * Allows you to set the direction of the counts and velocity without modifying the motor's direction state
+     * @param direction either reverse or forward depending on if encoder counts should be negated
+     */
+    public void setDirection(Direction direction) {
+        this.direction = direction;
+    }
+
+    public int getCurrentPosition() {
+        int multiplier = direction.getMultiplier();
+        int currentPosition = motor.getCurrentPosition() * multiplier;
+        if (currentPosition != lastPosition) {
+            double currentTime = clock.seconds();
+            double dt = currentTime - lastUpdateTime;
+            velocityEstimate = (currentPosition - lastPosition) / dt;
+            lastPosition = currentPosition;
+            lastUpdateTime = currentTime;
+        }
+        return currentPosition;
+    }
+
+    public double getVelocity() {
```

Comment:
I would qualify the name so it's obvious that this is compensating for overflow and doesn't match the behavior of `DcMotorEx#getVelocity()`. Having a `getVelocity()` that doesn't compensate would be nice.